### ISSUE TITLE
[FR] Safe softmax

### ISSUE THREAD

**Issue by jbschlosser** [pytorch/pytorch#55056]
## 🚀 Feature
Expand softmax to support "safe softmax" behavior (i.e. output 0 instead of NaN when called with no inputs).

## Motivation
When softmax is called with no inputs, the output is NaN due to computation of 0/0. In certain cases, an output of 0 may be preferred instead.

For example, due to masking in MHA, a batch may include a sentence that is "masked out" / consists entirely of padding. While this case will currently result in NaNs, leading to training divergence, it is often preferred to effectively ignore these cases so training can continue. See https://github.com/pytorch/pytorch/issues/41508 and https://github.com/pytorch/pytorch/issues/40932 for more info.

## Pitch
Add a new, optional `eps` parameter with default `None` to the various softmax forms:

```python
from torch import nn
from torch.nn import functional as F

m = nn.Softmax(dim=-1, eps=1e-5)
m2 = nn.Softmax2d(eps=1e-5)
output = F.softmax(..., eps=1e-5)
```

If specified, `eps` will be included in the denominator of the softmax computation, resulting in `0/(0+eps) = 0` instead of `0/0 = NaN` for the no input case. See https://github.com/pytorch/pytorch/issues/41508#issuecomment-775030992.

## Alternatives
Leave as-is; calling softmax with no inputs will unconditionally result in NaNs. It is left to users to ensure this doesn't happen if it's undesired, and failure to handle this correctly will still be considered "user error".

**Comment by jbschlosser** [comment]
Update: The various underlying softmax kernels use a common trick to improve numerical stability - subtract the max value before taking `exp()`. For the all `-inf` case, this gives NaNs:

```python
>>> float('-inf') - float('-inf')
nan
```

So adding an `eps` param doesn't work to get the desired all-zero output for an all `-inf` input.

One alternative is to check for `max value == -inf` and explicitly return 0s in that case. For this, I think a well-named flag that turns on the check makes more sense than an `eps` arg. As the max value is already being computed, it ends up being fine performance-wise to identify the all `-inf` case from within the softmax kernels.

**Comment by jbschlosser** [comment]
Update: Rather than implement "safe softmax" with its niche use case, it would be more flexible to implement masking behavior for softmax, where masking out an entire row results in 0s.

One possibility is to use sparse tensors for masks and implement sparse softmax kernels with the above semantics. The [Sputnik kernels](https://github.com/google-research/sputnik/blob/master/sputnik/softmax/sparse_softmax.cu.cc) are a candidate to be adapted for this purpose.

**Comment by mruberry** [comment]
cc @cpuhrsch

**Comment by thomasjpfan** [comment]
> Update: Rather than implement "safe softmax" with its niche use case, it would be more flexible to implement masking behavior for softmax, where masking out an entire row results in 0s.

For the MHA usecase, one would still need to compute the mask to pass into `softmax`, which means scanning through all of `attn_mask`. If we are okay this additional overhead, we can currently scan over the softmax result and zero out the rows that are completely masked.

> One alternative is to check for max value == -inf and explicitly return 0s in that case. For this, I think a well-named flag that turns on the check makes more sense than an eps arg. As the max value is already being computed, it ends up being fine performance-wise to identify the all -inf case from within the softmax kernels.

`zero_if_all_neg_inf`? If we want to make it more generic we can have: `value_for_nan` (default `nan`), where we set the value of softmax to this value when the max is `-inf`, `inf`, or `nan`. For the MHA usecase, `value_for_nan=0`.

**Comment by jbschlosser** [comment]
> For the MHA usecase, one would still need to compute the mask to pass into `softmax`, which means scanning through all of `attn_mask`. If we are okay this additional overhead, we can currently scan over the softmax result and zero out the rows that are completely masked.

Hm, I was thinking that `attn_mask` could be used directly in the softmax call, but you may be right since `attn_mask` in MHA can be either 2D or 3D.

> `zero_if_all_neg_inf`? If we want to make it more generic we can have: `value_for_nan` (default `nan`), where we set the value of softmax to this value when the max is `-inf`, `inf`, or `nan`. For the MHA usecase, `value_for_nan=0`.

I think we want the naming to indicate that this only applies for the NaN that comes about when `max_value == -inf` for a given row. There may be other NaN-producing cases that we should leave as-is (e.g. the `+inf` case shouldn't be zeroed). `zero_if_all_neg_inf` is pretty explicit so not terrible :)
